sympy/parsing/mathematica.py

```diff
@@ -42,6 +46,10 @@ def parse(s):
         (r"\A(-? *[\d\.]+)([a-zA-Z].*)\Z",  # Implied multiplication - 2a
         lambda m: parse(m.group(1)) + "*" + parse(m.group(2))),
 
+        # Multiplication between any object and a function
+        (r"\A(.+)([*])(\w+\[[^\]]+[^\[]*\])\Z",
```

Comment:
Suppose I use pattern
```python 
r"\A(.+)([ ])(\w+\[[^\]]+[^\[]*\])\Z"
```
Then it will also match functions such as `x + Sin[x]` (into `x + ` and `Sin[x]`). I think we need an algorithm to detect when `space` imply `multiplication`.